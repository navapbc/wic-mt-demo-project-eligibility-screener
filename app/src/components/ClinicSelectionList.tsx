import clinics from '@public/clinic-output/clinics-with-ids.json'
import { Trans } from 'next-i18next'
import {
  ChangeEvent,
  Dispatch,
  MouseEvent,
  ReactElement,
  SetStateAction,
} from 'react'

import Button from '@components/Button'
import ButtonLink, { ButtonLinkProps } from '@components/ButtonLink'
import ClinicInfo from '@components/ClinicInfo'
import Required from '@components/Required'

type ClinicSelectionListProps = {
  filteredClinics: (typeof clinics[0] | undefined)[]
  expandList: boolean
  setExpandList: Dispatch<SetStateAction<boolean>> | (() => unknown)
  selectedClinic: typeof clinics[0] | undefined
  handleSelection: (e: ChangeEvent<HTMLInputElement>) => void
  disabled: boolean
  continueBtn: ButtonLinkProps
}

const ClinicSelectionList = (props: ClinicSelectionListProps): ReactElement => {
  const {
    filteredClinics,
    expandList,
    setExpandList,
    selectedClinic,
    handleSelection,
    disabled,
    continueBtn,
  } = props

  const unexpandedNumberClinics = 4

  const handleClick = (e: MouseEvent<HTMLElement>) => {
    e.preventDefault()
    setExpandList(true)
  }

  let list: ReactElement = <></>
  if (filteredClinics.length > 0) {
    list = (
      <>
        <h2>
          <Trans i18nKey="ChooseClinic.listTitle" />
          <Required />
        </h2>
        <form className="usa-form usa-form--large">
          <fieldset className="usa-fieldset">
            {filteredClinics
              ?.slice(
                0,
                expandList ? filteredClinics.length : unexpandedNumberClinics
              )
              .map(
                (clinic, index) =>
                  clinic && (
                    <div className="usa-radio" key={index}>
                      <input
                        checked={selectedClinic?.clinic === clinic.clinic}
                        className="usa-radio__input usa-radio__input--tile"
                        id={clinic.clinic}
                        onChange={handleSelection}
                        type="radio"
                        value={clinic.clinic}
                      />
                      <label
                        className="usa-radio__label"
                        htmlFor={clinic.clinic}
                      >
                        <ClinicInfo
                          name={clinic.clinic}
                          streetAddress={clinic.clinicAddress}
                          phone={clinic.clinicTelephone}
                          isFormElement={true}
                        />
                      </label>
                    </div>
                  )
              )}
            {filteredClinics.length > 1 && !expandList && (
              <Button
                labelKey="ChooseClinic.showMoreOptions"
                style="unstyled"
                onClick={handleClick}
              />
            )}
          </fieldset>
          <ButtonLink
            disabled={disabled}
            href={continueBtn.href}
            labelKey={continueBtn.labelKey}
          />
        </form>
      </>
    )
  }

  return list
}

export default ClinicSelectionList
