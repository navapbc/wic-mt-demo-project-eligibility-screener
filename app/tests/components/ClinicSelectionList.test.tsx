/**
 * Test setup
 */
// Import the mocked list of filtered clinics.
// This path is mocked in jest.config.js to point at __mocks__.
import mockFilteredClinics from '@public/clinic-output/clinics-with-ids.json'
import { render, screen } from '@testing-library/react'
import cloneDeep from 'lodash/cloneDeep'

import ClinicSelectionList from '@components/ClinicSelectionList'

import { getMockClinic, setupClinicMocks } from '../helpers/setupClinics'
import { testSnapshot } from '../helpers/sharedTests'

setupClinicMocks()
const mockSelectedClinic = getMockClinic()
const props = {
  filteredClinics: mockFilteredClinics,
  expandList: false,
  setExpandList: jest.fn(),
  selectedClinic: mockSelectedClinic,
  handleSelection: jest.fn(),
  disabled: false,
  continueBtn: {
    labelKey: 'continue',
    route: '/somewhere',
  },
}

/**
 * Begin tests
 */
it('should match snapshot if the list is not expanded', () => {
  const testProps = cloneDeep(props)
  testSnapshot(<ClinicSelectionList {...testProps} />)
})

it('should match snapshot if the list is expanded', () => {
  const testProps = cloneDeep(props)
  testProps.expandList = true
  testSnapshot(<ClinicSelectionList {...testProps} />)
})

it('should match snapshot if the list is empty', () => {
  const testProps = cloneDeep(props)
  testProps.filteredClinics = []
  testSnapshot(<ClinicSelectionList {...testProps} />)
})

it('should display the right number of clinics if the list is not expanded', () => {
  const testProps = cloneDeep(props)
  render(<ClinicSelectionList {...testProps} />)
  const radioButtons = screen.getAllByRole('radio')
  expect(radioButtons.length).toBe(4)
})

it('should display the right number of clinics if the list is expanded', () => {
  const testProps = cloneDeep(props)
  testProps.expandList = true
  render(<ClinicSelectionList {...testProps} />)
  const radioButtons = screen.getAllByRole('radio')
  expect(radioButtons.length).toBe(mockFilteredClinics.length)
})

it('should not display the show more button if the list is already expanded', () => {
  const testProps = cloneDeep(props)
  testProps.expandList = true
  render(<ClinicSelectionList {...testProps} />)
  const button = screen.queryByRole('button', { name: /show more/i })
  expect(button).not.toBeInTheDocument()
})

it('should not display the show more button if the filterd clinics list has only one clinic', () => {
  const testProps = cloneDeep(props)
  testProps.filteredClinics = [getMockClinic()]
  render(<ClinicSelectionList {...testProps} />)
  const button = screen.queryByRole('button', { name: /show more/i })
  expect(button).not.toBeInTheDocument()
})

it('should only display the show more button if the filtered clinics list has multiple and the list is not already expanded', () => {
  const testProps = cloneDeep(props)
  render(<ClinicSelectionList {...testProps} />)
  const button = screen.getByRole('button', { name: /show more/i })
  expect(button).toBeInTheDocument()
})

it('should display the selected clinic as selected', () => {
  const testProps = cloneDeep(props)
  testProps.expandList = true
  render(<ClinicSelectionList {...testProps} />)
  const radioButton = screen.getByRole('radio', {
    name: new RegExp(mockSelectedClinic.clinic),
  })
  expect(radioButton).toBeChecked()
})

it('should control the action button', () => {
  const testProps = cloneDeep(props)
  testProps.disabled = true
  render(<ClinicSelectionList {...testProps} />)
  const button = screen.getByRole('button', { name: /continue/i })
  expect(button).toBeInTheDocument()
  expect(button).toBeDisabled()
})
