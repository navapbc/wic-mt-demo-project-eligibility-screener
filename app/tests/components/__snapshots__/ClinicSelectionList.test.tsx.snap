// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot if the list is empty 1`] = `<div />`;

exports[`should match snapshot if the list is expanded 1`] = `
<div>
  <h2>
    Choose a clinic from the following list:
    <abbr
      class="usa-hint usa-hint--required"
    >
       *
    </abbr>
  </h2>
  <form
    class="usa-form usa-form--large"
  >
    <fieldset
      class="usa-fieldset"
    >
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC ZERO"
          type="radio"
          value="CLINIC ZERO"
        />
        <label
          class="usa-radio__label"
          for="CLINIC ZERO"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC ZERO
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                0000 St, Helena, MT 00000
              </div>
              <div
                class="clinic-phone"
              >
                (000) 000-0000
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC ONE"
          type="radio"
          value="CLINIC ONE"
        />
        <label
          class="usa-radio__label"
          for="CLINIC ONE"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC ONE
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                1111 St, Helena, MT 11111
              </div>
              <div
                class="clinic-phone"
              >
                (111) 111-1111
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC TWO"
          type="radio"
          value="CLINIC TWO"
        />
        <label
          class="usa-radio__label"
          for="CLINIC TWO"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC TWO
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                2222 St, Helena, MT 22222
              </div>
              <div
                class="clinic-phone"
              >
                (222) 222-2222
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC THREE"
          type="radio"
          value="CLINIC THREE"
        />
        <label
          class="usa-radio__label"
          for="CLINIC THREE"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC THREE
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                3333 St, Helena, MT 33333
              </div>
              <div
                class="clinic-phone"
              >
                (333) 333-3333
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC FOUR"
          type="radio"
          value="CLINIC FOUR"
        />
        <label
          class="usa-radio__label"
          for="CLINIC FOUR"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC FOUR
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                4444 St, Helena, MT 44444
              </div>
              <div
                class="clinic-phone"
              >
                (444) 444-4444
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC FIVE"
          type="radio"
          value="CLINIC FIVE"
        />
        <label
          class="usa-radio__label"
          for="CLINIC FIVE"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC FIVE
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                5555 St, Helena, MT 55555
              </div>
              <div
                class="clinic-phone"
              >
                (555) 555-5555
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC SIX"
          type="radio"
          value="CLINIC SIX"
        />
        <label
          class="usa-radio__label"
          for="CLINIC SIX"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC SIX
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                6666 St, Helena, MT 66666
              </div>
              <div
                class="clinic-phone"
              >
                (666) 666-6666
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC SEVEN"
          type="radio"
          value="CLINIC SEVEN"
        />
        <label
          class="usa-radio__label"
          for="CLINIC SEVEN"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC SEVEN
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                7777 St, Helena, MT 77777
              </div>
              <div
                class="clinic-phone"
              >
                (777) 777-7777
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          checked=""
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC EIGHT"
          type="radio"
          value="CLINIC EIGHT"
        />
        <label
          class="usa-radio__label"
          for="CLINIC EIGHT"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC EIGHT
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                8888 St, Helena, MT 88888
              </div>
              <div
                class="clinic-phone"
              >
                (888) 888-8888
              </div>
            </div>
          </div>
        </label>
      </div>
    </fieldset>
    <button
      class="usa-button usa-button--small display-block margin-top-6"
    >
      Continue
    </button>
  </form>
</div>
`;

exports[`should match snapshot if the list is not expanded 1`] = `
<div>
  <h2>
    Choose a clinic from the following list:
    <abbr
      class="usa-hint usa-hint--required"
    >
       *
    </abbr>
  </h2>
  <form
    class="usa-form usa-form--large"
  >
    <fieldset
      class="usa-fieldset"
    >
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC ZERO"
          type="radio"
          value="CLINIC ZERO"
        />
        <label
          class="usa-radio__label"
          for="CLINIC ZERO"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC ZERO
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                0000 St, Helena, MT 00000
              </div>
              <div
                class="clinic-phone"
              >
                (000) 000-0000
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC ONE"
          type="radio"
          value="CLINIC ONE"
        />
        <label
          class="usa-radio__label"
          for="CLINIC ONE"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC ONE
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                1111 St, Helena, MT 11111
              </div>
              <div
                class="clinic-phone"
              >
                (111) 111-1111
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC TWO"
          type="radio"
          value="CLINIC TWO"
        />
        <label
          class="usa-radio__label"
          for="CLINIC TWO"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC TWO
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                2222 St, Helena, MT 22222
              </div>
              <div
                class="clinic-phone"
              >
                (222) 222-2222
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="CLINIC THREE"
          type="radio"
          value="CLINIC THREE"
        />
        <label
          class="usa-radio__label"
          for="CLINIC THREE"
        >
          <div
            class="clinic-info"
          >
            <div
              class="clinic-name"
            >
              CLINIC THREE
            </div>
            <div
              class="usa-checkbox__label-description"
            >
              <div
                class="clinic-street-address"
              >
                3333 St, Helena, MT 33333
              </div>
              <div
                class="clinic-phone"
              >
                (333) 333-3333
              </div>
            </div>
          </div>
        </label>
      </div>
      <button
        class="usa-button usa-button--small display-block usa-button--unstyled margin-top-1"
      >
        Show more clinic options
      </button>
    </fieldset>
    <button
      class="usa-button usa-button--small display-block margin-top-6"
    >
      Continue
    </button>
  </form>
</div>
`;
